File: app/static/js/country_editor.js
class CountryEditor {
    constructor() {
        this.modal = null;
    }

    init() {
        this.createUI();
    }

    createUI() {
        // Add country editor button to sidebar - ONLY IF NOT EXISTS
        if (!$('#create-country-btn').length) {
            const countryButton = $(`
                <button class="btn btn-info w-100 mb-2" id="create-country-btn">
                    <i class="bi bi-flag me-2"></i>Create Country
                </button>
            `);
            
            $('#open-project-btn').after(countryButton);
            
            // Create modal for country creation
            this.createModal();
            
            // Bind event listeners
            $('#create-country-btn').on('click', () => this.openCountryCreator());
            $('#confirm-country-create').on('click', () => this.createCountry());
            $('#country-color-picker').on('input', (e) => this.updateColorPreview(e.target.value));
        }
    }

    createModal() {
        // Only create modal if it doesn't exist
        if (!$('#country-creator-modal').length) {
            const modalHTML = `
            <div class="modal fade" id="country-creator-modal" tabindex="-1">
                <div class="modal-dialog">
                    <div class="modal-content bg-dark text-light">
                        <div class="modal-header border-secondary">
                            <h5 class="modal-title"><i class="bi bi-flag me-2"></i>Create New Country</h5>
                            <button type="button" class="btn-close btn-close-white" data-bs-dismiss="modal"></button>
                        </div>
                        <div class="modal-body">
                            <div class="mb-3">
                                <label class="form-label">Country Tag</label>
                                <input type="text" class="form-control bg-dark text-light border-secondary" 
                                       id="country-tag" maxlength="3" placeholder="e.g., USA, GER, SOV">
                                <div class="form-text text-muted">3 uppercase letters</div>
                            </div>
                            
                            <div class="mb-3">
                                <label class="form-label">Country Name</label>
                                <input type="text" class="form-control bg-dark text-light border-secondary" 
                                       id="country-name" placeholder="e.g., United States, Germany, Soviet Union">
                            </div>
                            
                            <div class="mb-3">
                                <label class="form-label">Color</label>
                                <div class="input-group">
                                    <input type="color" class="form-control form-control-color bg-dark border-secondary" 
                                           id="country-color-picker" value="#3d85c6">
                                    <span class="input-group-text bg-dark border-secondary">
                                        <div id="color-preview" style="width: 20px; height: 20px; background-color: #3d85c6;"></div>
                                    </span>
                                </div>
                            </div>
                            
                            <div class="mb-3">
                                <label class="form-label">Graphical Culture</label>
                                <select class="form-select bg-dark text-light border-secondary" id="graphical-culture">
                                    <option value="western_european_gfx">Western European</option>
                                    <option value="eastern_european_gfx">Eastern European</option>
                                    <option value="middle_eastern_gfx">Middle Eastern</option>
                                    <option value="asian_gfx">Asian</option>
                                    <option value="south_american_gfx">South American</option>
                                    <option value="african_gfx">African</option>
                                    <option value="neutral_gfx">Neutral</option>
                                    <option value="generic_gfx">Generic</option>
                                </select>
                            </div>
                            
                            <div class="mb-3">
                                <label class="form-label">2D Graphical Culture</label>
                                <select class="form-select bg-dark text-light border-secondary" id="graphical-culture-2d">
                                    <option value="western_european_2d">Western European 2D</option>
                                    <option value="eastern_european_2d">Eastern European 2D</option>
                                    <option value="middle_eastern_2d">Middle Eastern 2D</option>
                                    <option value="asian_2d">Asian 2D</option>
                                    <option value="south_american_2d">South American 2D</option>
                                    <option value="african_2d">African 2D</option>
                                    <option value="neutral_2d">Neutral 2D</option>
                                    <option value="generic_2d">Generic 2D</option>
                                </select>
                            </div>
                            
                            <div class="alert alert-info">
                                <small>
                                    <i class="bi bi-info-circle me-1"></i>
                                    This will create a country file in <code>common/countries/</code> and 
                                    add the tag to <code>common/country_tags/00_countries.txt</code>
                                </small>
                            </div>
                        </div>
                        <div class="modal-footer border-secondary">
                            <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Cancel</button>
                            <button type="button" class="btn btn-primary" id="confirm-country-create">
                                <i class="bi bi-plus-circle me-1"></i>Create Country
                            </button>
                        </div>
                    </div>
                </div>
            </div>
            `;
            
            $('body').append(modalHTML);
            this.modal = new bootstrap.Modal(document.getElementById('country-creator-modal'));
        }
    }

    openCountryCreator() {
        if (!window.editor?.currentProject) {
            alert('Please open a project first!');
            return;
        }
        
        // Reset form
        $('#country-tag').val('').prop('disabled', false);
        $('#country-name').val('');
        $('#country-color-picker').val('#3d85c6');
        $('#graphical-culture').val('western_european_gfx');
        $('#graphical-culture-2d').val('western_european_2d');
        this.updateColorPreview('#3d85c6');
        
        this.modal.show();
    }

    updateColorPreview(color) {
        $('#color-preview').css('background-color', color);
    }

    async createCountry() {
        // Prevent double submission
        if (this.isCreating) {
            return;
        }

        const tag = $('#country-tag').val().toUpperCase().trim();
        const name = $('#country-name').val().trim();
        const color = $('#country-color-picker').val();
        const graphicalCulture = $('#graphical-culture').val();
        const graphicalCulture2d = $('#graphical-culture-2d').val();

        // Validation
        if (!tag || !name) {
            alert('Please fill in all fields');
            return;
        }

        if (!/^[A-Z]{3}$/.test(tag)) {
            alert('Country tag must be exactly 3 uppercase letters');
            return;
        }

        this.isCreating = true;
        $('#confirm-country-create').prop('disabled', true).html('<i class="bi bi-hourglass me-1"></i>Creating...');

        try {
            const response = await fetch('/api/create_country', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    tag: tag,
                    name: name,
                    color: color,
                    graphical_culture: graphicalCulture,
                    graphical_culture_2d: graphicalCulture2d
                })
            });

            const result = await response.json();
            
            if (result.success) {
                alert(`Success: ${result.message}`);
                this.modal.hide();
                
                // Refresh file tree to show new files
                if (window.editor.currentProject) {
                    // Re-open the project to refresh the file tree
                    const response = await fetch('/api/open_project', {
                        method: 'POST',
                        headers: { 'Content-Type': 'application/json' },
                        body: JSON.stringify({ path: window.editor.currentProject })
                    });
                    
                    const result = await response.json();
                    if (result.success) {
                        window.editor.renderFileTree(result.structure);
                    }
                }
            } else {
                alert(`Error: ${result.message}`);
            }
        } catch (error) {
            alert('Failed to create country: ' + error.message);
        } finally {
            // Re-enable the button
            this.isCreating = false;
            $('#confirm-country-create').prop('disabled', false).html('<i class="bi bi-plus-circle me-1"></i>Create Country');
        }
    }
}

File: editors/country_editor.py
import os
import re
from pathlib import Path

class CountryCreator:
    def __init__(self, project_root):
        self.project_root = project_root
        self.countries_dir = os.path.join(project_root, "common", "countries")
        self.tags_file = os.path.join(project_root, "common", "country_tags", "00_countries.txt")
        
        # Available graphical cultures (from HOI4)
        self.graphical_cultures = [
            "western_european_gfx", "western_european_2d",
            "eastern_european_gfx", "eastern_european_2d", 
            "middle_eastern_gfx", "middle_eastern_2d",
            "asian_gfx", "asian_2d",
            "south_american_gfx", "south_american_2d",
            "african_gfx", "african_2d",
            "neutral_gfx", "neutral_2d",
            "generic_gfx", "generic_2d"
        ]
    
    def create_country(self, tag, name, color, graphical_culture, graphical_culture_2d=None):
        """Create a new country with given parameters"""
        if not graphical_culture_2d:
            graphical_culture_2d = graphical_culture.replace('_gfx', '_2d')
        
        # Validate tag
        if not re.match(r'^[A-Z]{3}$', tag):
            return False, "Tag must be exactly 3 uppercase letters"
        
        # Create countries directory if it doesn't exist
        os.makedirs(self.countries_dir, exist_ok=True)
        
        # Create country file
        country_file = os.path.join(self.countries_dir, f"{name.replace(' ', '_')}.txt")
        country_content = self._generate_country_file(color, graphical_culture, graphical_culture_2d)
        
        try:
            with open(country_file, 'w', encoding='utf-8') as f:
                f.write(country_content)
        except Exception as e:
            return False, f"Failed to create country file: {str(e)}"
        
        # Add to country tags
        tags_dir = os.path.dirname(self.tags_file)
        os.makedirs(tags_dir, exist_ok=True)
        
        tag_entry = f'{tag} = "countries/{os.path.basename(country_file)}"\n'
        
        try:
            # Read existing tags if file exists
            existing_tags = ""
            if os.path.exists(self.tags_file):
                with open(self.tags_file, 'r', encoding='utf-8') as f:
                    existing_tags = f.read()
            
            # Check if tag already exists
            if f'{tag} = ' in existing_tags:
                return False, f"Tag {tag} already exists in 00_countries.txt"
            
            # Append new tag
            with open(self.tags_file, 'a', encoding='utf-8') as f:
                f.write(tag_entry)
                
        except Exception as e:
            return False, f"Failed to update country tags: {str(e)}"
        
        return True, f"Country {name} ({tag}) created successfully!"
    
    def _generate_country_file(self, color, graphical_culture, graphical_culture_2d):
        """Generate the country file content"""
        r, g, b = color
        return f"""graphical_culture = {graphical_culture}
graphical_culture_2d = {graphical_culture_2d}

color = {{ {r} {g} {b} }}
"""
    
    def get_available_graphical_cultures(self):
        """Get list of available graphical cultures"""
        return [gc for gc in self.graphical_cultures if gc.endswith('_gfx')]
    
    def validate_color(self, color_string):
        """Validate and parse color string (can be RGB tuple or hex)"""
        try:
            if isinstance(color_string, str) and color_string.startswith('#'):
                # Hex color
                hex_color = color_string.lstrip('#')
                if len(hex_color) == 6:
                    r = int(hex_color[0:2], 16)
                    g = int(hex_color[2:4], 16) 
                    b = int(hex_color[4:6], 16)
                    return (r, g, b)
            elif isinstance(color_string, (list, tuple)) and len(color_string) == 3:
                # RGB tuple
                return tuple(int(c) for c in color_string)
        except:
            pass
        return None
    
    def get_existing_tags(self):
        """Get list of existing country tags"""
        if not os.path.exists(self.tags_file):
            return []
        
        try:
            with open(self.tags_file, 'r', encoding='utf-8') as f:
                content = f.read()
            
            # Extract tags using regex
            tags = re.findall(r'^(\w{3})\s*=', content, re.MULTILINE)
            return tags
        except:
            return []
